packages/data-lake-api/src/DataLakeEventsFetchQuery.js

```diff
@@ -0,0 +1,104 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { SqlQuery } from './SqlQuery';
+
+export class DataLakeEventsFetchQuery {
+  constructor(database, options) {
+    this.database = database;
+
+    const {
+      organisationUnitCodes,
+      startDate,
+      endDate,
+      eventId,
+      dataGroupCode,
+      dataElementCodes,
+    } = options;
+
+    this.dataElementCodes = dataElementCodes;
+    this.entityCodes = organisationUnitCodes;
+    this.startDate = startDate;
+    this.endDate = endDate;
+    this.eventId = eventId;
+    this.dataGroupCode = dataGroupCode;
+    this.hasDataElements = dataElementCodes && dataElementCodes.length > 0;
+  }
+
+  async fetch() {
+    const { query, params } = this.buildQueryAndParams();
+
+    const sqlQuery = new SqlQuery(query, params);
+
+    return sqlQuery.executeOnDatabase(this.database);
+  }
+
+  getAliasedColumns() {
+    const aliasedColumns = [
+      'date',
+      'entity_code AS "entityCode"',
+      'entity_code AS "entityName"',
+      'event_id as "eventId"',
+      `'Number' as "type"`,
+      'value',
+    ];
+    if (this.hasDataElements) {
+      aliasedColumns.push('data_element_code AS "dataElementCode"');
+    }
+    return aliasedColumns.join(', ');
+  }
+
+  getInnerJoinsAndParams() {
+    let params = this.entityCodes;
+    const joins = [SqlQuery.innerJoin('analytics', 'entity_code', this.entityCodes)];
```

Comment:
How come this query has joins on the same table?

Reply:
Yeah the joins here a bit a surprising, but it's a performance optimisation we came across when noticing slowness in the analytics table.

Basically what we want to be saying here is: `SELECT * from analytics WHERE entity_code IN ${this.entityCodes}`
However, when there are looots of entity codes (which happens sometimes with large hierarchies, ie. all schools in Laos) this can be quite slow.

The recommendation seems to be to instead use a `VALUES` expression and join it as so:
```
SELECT * FROM analytics
INNER JOIN (
  VALUES ${this.entityCodes}
) entity_codes(code) on entity_codes.code = analytics.entity_code
```

https://dba.stackexchange.com/questions/91247/optimizing-a-postgres-query-with-a-large-in

Reply:
Ah cool, can we pop that in as a comment, super cool trick